management-node/server.go

```diff
@@ -0,0 +1,530 @@
+			Method: common.WorkerNodeRPCExec,
+			Params: params,
+		}
+
+		// task is scheduled
+		// setup dead timeout
+		md.SetTaskAndRunDeadTimeout(&Task{
+			task: task,
+			stop: make(chan struct{}, 1),
+		})
+
+		// sending to worker node
+		md.workerNodePool[workerId].
+			SendWithHandlerTimeout(req, onRspHandler,
+				onTimerExpiredHandler,
+				time.Duration(Config.SilenceTimeout))
+	}
+
+	// prepare task objects to response to schedctl
+	tasks := []*pb.Task{}
+	for _ = range req.Tasks {
+		m := <-messanger
+
+		// check if it responded with error
+		if m.err != nil {
+			log.Error(m.err)
+
+			// set error to worker error field
+			m.task.WorkerError = m.err.Error()
+			if err := md.SetToDb(m.task, EtcdTaskPrefix+m.task.Id); err != nil {
+				log.Error(err.Error())
+				return nil, err
+			}
+			tasks = append(tasks, m.task)
+			continue
+		}
+
+		tasks = append(tasks, m.task)
+	}
+
+	close(messanger)
+
+	return &pb.TaskList{Tasks: tasks}, nil
+
+}
+
+// SetToDb stores subj to etcd by id
+func (md *ManagementNode) SetToDb(subj interface{}, id string) error {
+	b, err := json.Marshal(subj)
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(),
+		time.Duration(Config.DialTimeout)*time.Second)
+	_, err = md.etcd.Put(ctx, id, string(b))
+	cancel()
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+
+	return nil
+}
+
+// SetTaskState is gRPC that is called by worker node
+// to set state of the task that was previously executed
+// it stops dead timeout
+// and sets task to db
+func (md *ManagementNode) SetTaskState(ctx context.Context, task *pb.Task) (*pb.Empty, error) {
+
+	// TBD: check of existed object might be needed
+	// might be dead before state change
+
+	log.Infof("SetTaskState with params %+v", task)
+
+	if err := md.StopTaskDeadTimeout(task.Id); err != nil {
+		log.Error(err.Error())
+		return nil, err
+	}
+
+	if err := md.SetToDb(task, EtcdTaskPrefix+task.Id); err != nil {
+		log.Error(err.Error())
+		return nil, err
+	}
+
+	return &pb.Empty{}, nil
+}
+
+// setTasksToDb is helper function for setting bunch of tasks
+// to etcd
+func (md *ManagementNode) setTasksToDb(tasks []*pb.Task) error {
+	for _, task := range tasks {
+		err := md.SetToDb(task, EtcdTaskPrefix+task.Id)
+		if err != nil {
+			log.Error(err.Error())
+			return err
+		}
+	}
+	return nil
+}
+
+// PrepareZeroState sets proper state for workers and tasks
+// in db
+// workers are all set to disconnected
+// task with pending and scheduled are all set to dead
+func (md *ManagementNode) PrepareZeroState() error {
+
+	tl, err := md.GetTaskList(context.Background(), &pb.DummyReq{})
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+
+	for _, task := range tl.Tasks {
+		if task.State != common.TaskStateDone && task.State != common.TaskStateDead {
+			task.State = common.TaskStateDead
+			err := md.SetToDb(task, EtcdTaskPrefix+task.Id)
+			if err != nil {
+				log.Error(err.Error())
```

Comment:
Just to reiterate something I mentioned above, instead of logging each error every time, just propagate them to the top level and log there, so all `log.Error(err.Error())` aren't necessary (and use `trace.Wrap(err)`)

Reply:
done